common/src/main/java/com/sequenceiq/cloudbreak/certificate/PkiUtil.java

```diff
@@ -118,15 +137,22 @@ public static String generateSignature(String privateKeyPem, byte[] data) {
                 RSAPrivateKeySpec privKeySpec = factory.getKeySpec(kp.getPrivate(), RSAPrivateKeySpec.class);
                 rsaKeyParameters = new RSAKeyParameters(true, privKeySpec.getModulus(), privKeySpec.getPrivateExponent());
 
-                CACHE.put(privateKeyPem, rsaKeyParameters);
+                RSA_KEY_PARAMS_CACHE.put(privateKeyPem, rsaKeyParameters);
             } catch (NoSuchAlgorithmException | IOException | InvalidKeySpecException e) {
                 throw new SecurityException(e);
             }
         }
 
-        Signer signer = new PSSSigner(new RSAEngine(), new SHA256Digest(), SALT_LENGTH);
-        signer.init(true, rsaKeyParameters);
-        signer.update(data, 0, data.length);
+        Signer signer = PSS_SIGNER_CACHE.get(rsaKeyParameters);
+
+        if (signer == null) {
+            signer = new PSSSigner(new RSAEngine(), new SHA256Digest(), SALT_LENGTH);
+            signer.init(true, rsaKeyParameters);
+            signer.update(data, 0, data.length);
+
+            PSS_SIGNER_CACHE.put(rsaKeyParameters, signer);
```

Comment:
little bit looks weird. The map is static final but modifies the content of the map. Why is it static final?

Reply:
the reference is final and I can initialize it in static initializer block

Reply:
I understand that, but why do you create the map as static final?

Reply:
1) because if it is not static then I can not initialize it in static block
2) It is static because I wanted one instance per JVM
3) I followed the previous CACHE solution :)